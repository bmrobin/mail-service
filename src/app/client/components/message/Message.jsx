import React  from 'react';
import { Form, FormGroup, FormControl, ControlLabel, Col, Button } from 'react-bootstrap';
import { sendEmail } from 'utils/api';
import Error from 'components/error/Error';

export default class Message extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  sendEmail(event) {
    event.preventDefault();
    console.log("Sending email to contact list...", JSON.stringify(this.state));
    sendEmail(this.state.message).then(() => {
      console.log('success!');
      this.setState({ message: undefined });
      console.log(this.state);
    }).catch(() => {
      console.error('error!');
      this.setState({ connectionError: true });
    });
  }

  handler(event) {
    this.setState({
      message: event.target.value,
      connectionError: undefined
    });
  }

  render() {
    return (
      <Form horizontal onSubmit={(e) => this.sendEmail(e)}>

        <FormGroup>
          <Col sm={2} componentClass={ControlLabel}>
            Message
          </Col>
          <Col sm={10}>
            <FormControl componentClass="textarea" id="message-id" onChange={(e) => this.handler(e)} />
          </Col>
        </FormGroup>


        <Button type="submit" 
                className="save-btn"
                id="save-id"
                onSubmit={(event) => this.sendEmail(event)}>
          Send
        </Button>

        {this.state.connectionError && <Error message="An error occurred while sending message." />}

      </Form>
    );
  }
}
