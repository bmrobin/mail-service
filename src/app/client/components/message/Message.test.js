import React from 'react';
import { shallow } from 'enzyme';
import { FormControl, Button } from 'react-bootstrap';
import Message from './Message';

jest.mock('utils/api', () => (
  {
    sendEmail: jest.fn((args) => {
      return new Promise((resolve, reject) => {
        if (args === 'email') {
          resolve();
        } else {
          reject();
        }
      });
    })
  }
));

describe("Message Component", () => {
  test("should render component", () => {
    const component = shallow(<Message />);
    expect(component.debug()).toMatchSnapshot();
  });

  test("should store message when changed in form", () => {
    const component = shallow(<Message />);
    expect(component.state().message).toBeUndefined();
    component.find(FormControl).simulate('change', {target: { value: 'my value' }});
    expect(component.state().message).toBe('my value');
  });

  test.skip("should send an email when button is clicked", () => {
    const mockEventFn = jest.fn();
    const component = shallow(<Message />);
    component.find(FormControl).simulate('change', {target: { value: 'email' }});
    component.find(Button).simulate('submit', { preventDefault: mockEventFn });
    expect(mockEventFn).toHaveBeenCalledTimes(1);
    expect(component.state()).toEqual({ connectionError: undefined, message: undefined });
  });

});
