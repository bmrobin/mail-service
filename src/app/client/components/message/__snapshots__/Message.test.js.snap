// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Component should render component 1`] = `
"<Form horizontal={true} onSubmit={[Function: onSubmit]} inline={false} componentClass=\\"form\\" bsClass=\\"form\\">
  <FormGroup bsClass=\\"form-group\\">
    <Col sm={2} componentClass={[Function: ControlLabel]} bsClass=\\"col\\">
      Message
    </Col>
    <Col sm={10} componentClass=\\"div\\" bsClass=\\"col\\">
      <FormControl componentClass=\\"textarea\\" id=\\"message-id\\" onChange={[Function: onChange]} bsClass=\\"form-control\\" />
    </Col>
  </FormGroup>
  <Button type=\\"submit\\" className=\\"save-btn\\" id=\\"save-id\\" onSubmit={[Function: onSubmit]} active={false} block={false} disabled={false} bsStyle=\\"default\\" bsClass=\\"btn\\">
    Send
  </Button>
</Form>"
`;
